patches/server/0289-Restore-custom-InventoryHolder-support.patch

```diff
@@ -16,31 +16,284 @@ This provides a "best of both worlds" scenario: plugins with InventoryHolder's
 will always work as intended in the past, those without will create implementation
 based inventories.
 
-diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
-index 0899afd175f969da0df9371d96d3b5e1de4c8533..fc307213a4bc9675c6011a81e4e06cd23a22926d 100644
---- a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
-+++ b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
-@@ -40,6 +40,11 @@ public final class CraftInventoryCreator {
-     }
+diff --git a/src/main/java/io/papermc/paper/inventory/PaperInventoryCustom.java b/src/main/java/io/papermc/paper/inventory/PaperInventoryCustom.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..0918d08123f73a90f1c2b1afb7391ba5ca1bdd10
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/inventory/PaperInventoryCustom.java
+@@ -0,0 +1,137 @@
++package io.papermc.paper.inventory;
++
++import net.minecraft.world.Container;
++import net.minecraft.world.entity.player.Player;
++import net.minecraft.world.item.ItemStack;
++import org.bukkit.Location;
++import org.bukkit.craftbukkit.entity.CraftHumanEntity;
++import org.bukkit.craftbukkit.inventory.CraftInventory;
++import org.bukkit.entity.HumanEntity;
++import org.bukkit.event.inventory.InventoryType;
++import org.bukkit.inventory.InventoryHolder;
++
++import java.util.List;
++
++public class PaperInventoryCustom extends CraftInventory {
```

Comment:
I question the need for a whole new class just to do delegation of this. Why do we need it?

Reply:
To keep the container as the actual container from the tile entity, instead of the `MinecraftInventory` one. It should be a more "accurate" copy of the inventory that way.

Reply:
I don't quite get why that's an issue. What happens if you don't use a more "accurate" copy?

Reply:
I didn't test anything specific, but if a Container implementation has anything different in its setItem implementation or any of its method implementations this would catch it right? I see several differing implementation of setItem

Reply:
Yeah, that seems correct. If this fixes the issue and another @PaperMC/development-team member signs off with no better solution, I'm happy. I see now that it's an NMS container, so that should hopefully be fine...